src/layout.rs

```diff
@@ -559,6 +574,46 @@ impl Rect {
 mod tests {
     use super::*;
 
+    #[test]
+    fn test_horizontal_split_equally_in_underspecified_case() {
+        let target = Rect {
+            x: 2,
+            y: 2,
+            width: 10,
+            height: 10,
+        };
+
+        let chunks = Layout::default()
+            .direction(Direction::Horizontal)
+            .constraints([Constraint::Min(2), Constraint::Min(2), Constraint::Min(0)].as_ref())
+            .split(target);
+
+        assert_eq!(target.width, chunks.iter().map(|r| r.width).sum::<u16>());
+        assert_eq!(chunks[0].width, 3);
+        assert_eq!(chunks[1].width, 3);
+        assert_eq!(chunks[2].width, 4);
+    }
+
+    #[test]
+    fn test_vertical_split_equally_in_underspecified_case() {
+        let target = Rect {
+            x: 2,
+            y: 2,
+            width: 10,
+            height: 10,
+        };
+
+        let chunks = Layout::default()
+            .direction(Direction::Vertical)
+            .constraints([Constraint::Min(2), Constraint::Min(2), Constraint::Min(0)].as_ref())
+            .split(target);
+
+        assert_eq!(target.height, chunks.iter().map(|r| r.height).sum::<u16>());
+        assert_eq!(chunks[0].height, 3);
+        assert_eq!(chunks[1].height, 3);
+        assert_eq!(chunks[2].height, 4);
+    }
```

Comment:
What happens when this fails on a CI server somewhere because of some unrelated change (or due to something non-deterministic), meaning the failure is difficult to repro locally?

The first check failing here will prevent the 2nd to 4th checks executing, which contain the information you'd need to understand how to fix the test  A better test for this would show how those numbers change for various split widths:

```suggestion
    fn test_horizontal_split_equally_in_underspecified_case() {
        use Constraint::*;
        fn split_widths(constraints: &Vec<Constraint>, target_width: u16) -> Vec<u16> {
            let target = Rect::new(2, 2, target_width, 10);
            Layout::default()
                .direction(Direction::Horizontal)
                .constraints(constraints.as_ref())
                .split(target)
                .iter()
                .map(|r| r.width)
                .collect_vec()
        }

        let constraints = vec![Min(2), Min(2), Min(0)];
        assert_eq!(split_widths(&constraints, 0), vec![0, 0, 0]);
        assert_eq!(split_widths(&constraints, 1), vec![1, 0, 0]);
        assert_eq!(split_widths(&constraints, 2), vec![2, 0, 0]);
        assert_eq!(split_widths(&constraints, 3), vec![2, 1, 0]);
        assert_eq!(split_widths(&constraints, 4), vec![2, 2, 0]);
        assert_eq!(split_widths(&constraints, 5), vec![2, 2, 1]);
        assert_eq!(split_widths(&constraints, 6), vec![2, 2, 2]);
        assert_eq!(split_widths(&constraints, 7), vec![2, 2, 3]);
        assert_eq!(split_widths(&constraints, 8), vec![2, 2, 4]);
        assert_eq!(split_widths(&constraints, 9), vec![2, 2, 5]);
        assert_eq!(split_widths(&constraints, 10), vec![2, 2, 6]);
    }
```
Those are the numbers before the change to the split function is applied. Normally I much prefer to have tests be  linear and simpler than this one, but I think a good reason to find an exception is where the test communicates the relationship between the outputs of similar inputs that would be otherwise difficult to understand in isolation.

What this test doesn't communicate / test is where the edge cases on the new code take effect.

Oddly, I can't actually get this to succeed with the new numbers as this now fails non-deterministically. Is the code intended to be deterministic?

E.g. here are the results of `cargo test` run several times:
```
thread 'layout::tests::test_horizontal_split_equally_in_underspecified_case' panicked at 'assertion failed: `(left == right)`
  left: `[1, 1, 1]`,
 right: `[1, 1, 0]`', src/layout.rs:596:9
```
```
thread 'layout::tests::test_horizontal_split_equally_in_underspecified_case' panicked at 'assertion failed: `(left == right)`
  left: `[1, 0, 0]`,
 right: `[1, 1, 0]`', src/layout.rs:596:9
```
```
thread 'layout::tests::test_horizontal_split_equally_in_underspecified_case' panicked at 'assertion failed: `(left == right)`
  left: `[0, 0, 1]`,
 right: `[0, 0, 0]`', src/layout.rs:595:9
```
```
thread 'layout::tests::test_horizontal_split_equally_in_underspecified_case' panicked at 'assertion failed: `(left == right)`
  left: `[2, 2, 0]`,
 right: `[1, 1, 0]`', src/layout.rs:598:9
```

Reply:
This is a can of worms if it is non-deterministic. Maybe we should only have tests that are deterministic?

Reply:
Yes, we should only have deterministic tests, but so too behavior.

Ideally we need to find a way to push the constraints to choose more deterministic values for constraints - perhaps by being more granular with the weighting of the importance of each constraint? I think this probably becomes complex fast though, but the worms are there already I think.

Reply:
I ran it multiple times locally and it doesn't appear to be non-deterministic for me, although I'm not sure why since I thought the constraints were essentially the same as before.